File: src/lib/transcript.js
'use strict';

const fs = require('fs');

// Extract the last assistant text message from a Claude Code
// transcript JSONL file.
function getLastAssistantMessage(transcriptPath) {
  if (!transcriptPath || !fs.existsSync(transcriptPath)) {
    return null;
  }

  try {
    const content = fs.readFileSync(transcriptPath, 'utf8');
    const lines = content.trim().split('\n');

    for (let i = lines.length - 1; i >= 0; i--) {
      let entry;
      try {
        entry = JSON.parse(lines[i]);
      } catch (_e) {
        continue;
      }

      if (entry.type !== 'assistant') continue;

      const msg = entry.message;
      if (!msg || !msg.content || !Array.isArray(msg.content)) {
        continue;
      }

      let texts = [];
      let hasQuestion = false;

      for (const block of msg.content) {
        if (block.type === 'text' && block.text) {
          texts.push(block.text);
        } else if (
          block.type === 'tool_use' &&
          block.name === 'AskUserQuestion' &&
          block.input
        ) {
          hasQuestion = true;
          const questions = block.input.questions;
          if (Array.isArray(questions)) {
            for (const question of questions) {
              texts.push(question.question || '');
              if (Array.isArray(question.options)) {
                question.options.forEach((opt, idx) => {
                  texts.push(`${idx + 1}. ${opt.label}`);
                });
              }
            }
          }
        }
      }

      // If there's a question, prefer just the question content
      if (hasQuestion && texts.length > 1) {
        const questionStart = texts.findIndex(
          (t) => t.match(/\?($|\s)/) || t.match(/^\d+\./)
        );
        if (questionStart > 0) {
          texts = texts.slice(questionStart);
        }
      }

      if (texts.length > 0) {
        return texts.join('\n');
      }
    }
  } catch (_e) {
    // ignore read errors
  }

  return null;
}

// Returns { optionCount: N } if the last assistant message has
// an AskUserQuestion, or null if no question found.
function getLastQuestionMeta(transcriptPath) {
  if (!transcriptPath || !fs.existsSync(transcriptPath)) {
    return null;
  }

  try {
    const content = fs.readFileSync(transcriptPath, 'utf8');
    const lines = content.trim().split('\n');

    for (let i = lines.length - 1; i >= 0; i--) {
      let entry;
      try {
        entry = JSON.parse(lines[i]);
      } catch (_e) {
        continue;
      }

      if (entry.type !== 'assistant') continue;
      const msg = entry.message;
      if (!msg || !msg.content || !Array.isArray(msg.content)) {
        continue;
      }

      for (const block of msg.content) {
        if (
          block.type === 'tool_use' &&
          block.name === 'AskUserQuestion' &&
          block.input
        ) {
          const questions = block.input.questions;
          if (
            Array.isArray(questions) &&
            questions.length > 0 &&
            Array.isArray(questions[0].options)
          ) {
            return { optionCount: questions[0].options.length };
          }
        }
      }
      // Only check the last assistant message
      return null;
    }
  } catch (_e) {
    // ignore
  }
  return null;
}

module.exports = { getLastAssistantMessage, getLastQuestionMeta };

File: internal/transcript/transcript.go
package transcript

import (
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"strings"
)

var questionLineRegex = regexp.MustCompile(`(?:^\?($|\s)|^\d+\.)`)

type transcriptEntry struct {
	Type    string          `json:"type"`
	Message json.RawMessage `json:"message"`
}

type assistantMessage struct {
	Content []contentBlock `json:"content"`
}

type contentBlock struct {
	Type  string          `json:"type"`
	Text  string          `json:"text"`
	Name  string          `json:"name"`
	Input json.RawMessage `json:"input"`
}

type askUserInput struct {
	Questions []questionItem `json:"questions"`
}

type questionItem struct {
	Question string       `json:"question"`
	Options  []optionItem `json:"options"`
}

type optionItem struct {
	Label string `json:"label"`
}

type QuestionMeta struct {
	OptionCount int
}

func GetLastAssistantMessage(transcriptPath string) string {
	if transcriptPath == "" {
		return ""
	}
	data, err := os.ReadFile(transcriptPath)
	if err != nil {
		return ""
	}

	lines := strings.Split(strings.TrimSpace(string(data)), "\n")
	for i := len(lines) - 1; i >= 0; i-- {
		line := strings.TrimSpace(lines[i])
		if line == "" {
			continue
		}
		var entry transcriptEntry
		if err := json.Unmarshal([]byte(line), &entry); err != nil {
			continue
		}
		if entry.Type != "assistant" {
			continue
		}
		var msg assistantMessage
		if err := json.Unmarshal(entry.Message, &msg); err != nil {
			continue
		}

		var texts []string
		hasQuestion := false
		for _, block := range msg.Content {
			if block.Type == "text" && block.Text != "" {
				texts = append(texts, block.Text)
			}
			if block.Type == "tool_use" && block.Name == "AskUserQuestion" {
				var input askUserInput
				if err := json.Unmarshal(block.Input, &input); err == nil {
					for _, q := range input.Questions {
						if q.Question != "" {
							texts = append(texts, q.Question)
						}
						for idx, opt := range q.Options {
							texts = append(texts, fmt.Sprintf("%d. %s", idx+1, opt.Label))
						}
						if len(q.Options) > 0 {
							hasQuestion = true
						}
					}
				}
			}
		}

		if len(texts) == 0 {
			return ""
		}

		result := strings.Join(texts, "\n")
		if hasQuestion {
			resultLines := strings.Split(result, "\n")
			for i, line := range resultLines {
				if questionLineRegex.MatchString(line) {
					return strings.Join(resultLines[i:], "\n")
				}
			}
		}
		return result
	}
	return ""
}

func GetLastQuestionMeta(transcriptPath string) *QuestionMeta {
	if transcriptPath == "" {
		return nil
	}
	data, err := os.ReadFile(transcriptPath)
	if err != nil {
		return nil
	}

	lines := strings.Split(strings.TrimSpace(string(data)), "\n")
	for i := len(lines) - 1; i >= 0; i-- {
		line := strings.TrimSpace(lines[i])
		if line == "" {
			continue
		}
		var entry transcriptEntry
		if err := json.Unmarshal([]byte(line), &entry); err != nil {
			continue
		}
		if entry.Type != "assistant" {
			continue
		}
		var msg assistantMessage
		if err := json.Unmarshal(entry.Message, &msg); err != nil {
			continue
		}

		for _, block := range msg.Content {
			if block.Type == "tool_use" && block.Name == "AskUserQuestion" {
				var input askUserInput
				if err := json.Unmarshal(block.Input, &input); err == nil {
					for _, q := range input.Questions {
						if len(q.Options) > 0 {
							return &QuestionMeta{OptionCount: len(q.Options)}
						}
					}
				}
			}
		}
		return nil
	}
	return nil
}
